### Required Assignment 5.1: Will the Customer Accept the Coupon?

**Context**

Imagine driving through town and a coupon is delivered to your cell phone for a restaurant near where you are driving. Would you accept that coupon and take a short detour to the restaurant? Would you accept the coupon but use it on a subsequent trip? Would you ignore the coupon entirely? What if the coupon was for a bar instead of a restaurant? What about a coffee house? Would you accept a bar coupon with a minor passenger in the car? What about if it was just you and your partner in the car? Would weather impact the rate of acceptance? What about the time of day?

Obviously, proximity to the business is a factor on whether the coupon is delivered to the driver or not, but what are the factors that determine whether a driver accepts the coupon once it is delivered to them? How would you determine whether a driver is likely to accept a coupon?

**Overview**

The goal of this project is to use what you know about visualizations and probability distributions to distinguish between customers who accepted a driving coupon versus those that did not.

**Data**

This data comes to us from the UCI Machine Learning repository and was collected via a survey on Amazon Mechanical Turk. The survey describes different driving scenarios including the destination, current time, weather, passenger, etc., and then ask the person whether he will accept the coupon if he is the driver. Answers that the user will drive there ‘right away’ or ‘later before the coupon expires’ are labeled as ‘Y = 1’ and answers ‘no, I do not want the coupon’ are labeled as ‘Y = 0’.  There are five different types of coupons -- less expensive restaurants (under \$20), coffee houses, carry out & take away, bar, and more expensive restaurants (\$20 - $50).

**Deliverables**

Your final product should be a brief report that highlights the differences between customers who did and did not accept the coupons.  To explore the data you will utilize your knowledge of plotting, statistical summaries, and visualization using Python. You will publish your findings in a public facing github repository as your first portfolio piece.





### Data Description
Keep in mind that these values mentioned below are average values.

The attributes of this data set include:
1. User attributes
    -  Gender: male, female
    -  Age: below 21, 21 to 25, 26 to 30, etc.
    -  Marital Status: single, married partner, unmarried partner, or widowed
    -  Number of children: 0, 1, or more than 1
    -  Education: high school, bachelors degree, associates degree, or graduate degree
    -  Occupation: architecture & engineering, business & financial, etc.
    -  Annual income: less than \\$12500, \\$12500 - \\$24999, \\$25000 - \\$37499, etc.
    -  Number of times that he/she goes to a bar: 0, less than 1, 1 to 3, 4 to 8 or greater than 8
    -  Number of times that he/she buys takeaway food: 0, less than 1, 1 to 3, 4 to 8 or greater
    than 8
    -  Number of times that he/she goes to a coffee house: 0, less than 1, 1 to 3, 4 to 8 or
    greater than 8
    -  Number of times that he/she eats at a restaurant with average expense less than \\$20 per
    person: 0, less than 1, 1 to 3, 4 to 8 or greater than 8
    -  Number of times that he/she goes to a bar: 0, less than 1, 1 to 3, 4 to 8 or greater than 8
    

2. Contextual attributes
    - Driving destination: home, work, or no urgent destination
    - Location of user, coupon and destination: we provide a map to show the geographical
    location of the user, destination, and the venue, and we mark the distance between each
    two places with time of driving. The user can see whether the venue is in the same
    direction as the destination.
    - Weather: sunny, rainy, or snowy
    - Temperature: 30F, 55F, or 80F
    - Time: 10AM, 2PM, or 6PM
    - Passenger: alone, partner, kid(s), or friend(s)


3. Coupon attributes
    - time before it expires: 2 hours or one day

In [1]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np

### Problems

Use the prompts below to get started with your data analysis.  

1. Read in the `coupons.csv` file.




In [2]:
data = pd.read_csv('data/coupons.csv')

In [3]:
data.head()

,destination,passanger,weather,temperature,time,coupon,expiration,gender,age,maritalStatus,...,CoffeeHouse,CarryAway,RestaurantLessThan20,Restaurant20To50,toCoupon_GEQ5min,toCoupon_GEQ15min,toCoupon_GEQ25min,direction_same,direction_opp,Y
0,No Urgent Place,Alone,Sunny,55,2PM,Restaurant(<20),1d,Female,21,Unmarried partner,...,never,NaN,4~8,1~3,1,0,0,0,1,1
1,No Urgent Place,Friend(s),Sunny,80,10AM,Coffee House,2h,Female,21,Unmarried partner,...,never,NaN,4~8,1~3,1,0,0,0,1,0
2,No Urgent Place,Friend(s),Sunny,80,10AM,Carry out & Take away,2h,Female,21,Unmarried partner,...,never,NaN,4~8,1~3,1,1,0,0,1,1
3,No Urgent Place,Friend(s),Sunny,80,2PM,Coffee House,2h,Female,21,Unmarried partner,...,never,NaN,4~8,1~3,1,1,0,0,1,0
4,No Urgent Place,Friend(s),Sunny,80,2PM,Coffee House,1d,Female,21,Unmarried partner,...,never,NaN,4~8,1~3,1,1,0,0,1,0


2. Investigate the dataset for missing or problematic data.

In [4]:
# Investigate the dataset for missing or problematic data

# A. Initial data-quality investigation


# - data info

print("\033[1m" + "\nData info:\n" + "\033[0m")
data.info()


# - dataset dimensions

print("\033[1m" + "\nDataset size:\n" + "\033[0m")

rows = data.shape[0]
columns = data.shape[1]

print("Rows: " + str(rows) + "\nColumns: " + str(columns))


# - check unique values in categorical columns

categorical_columns = data.select_dtypes(exclude="number").columns

for column in categorical_columns:
    print("\033[1m" + "\n" + column + ":\n" + "\033[0m")
    for value in data[column].unique():
        print("•", value)


Data info:

<class 'pandas.DataFrame'>
RangeIndex: 12684 entries, 0 to 12683
Data columns (total 26 columns):
 #   Column                Non-Null Count  Dtype
---  ------                --------------  -----
 0   destination           12684 non-null  str  
 1   passanger             12684 non-null  str  
 2   weather               12684 non-null  str  
 3   temperature           12684 non-null  int64
 4   time                  12684 non-null  str  
 5   coupon                12684 non-null  str  
 6   expiration            12684 non-null  str  
 7   gender                12684 non-null  str  
 8   age                   12684 non-null  str  
 9   maritalStatus         12684 non-null  str  
 10  has_children          12684 non-null  int64
 11  education             12684 non-null  str  
 12  occupation            12684 non-null  str  
 13  income                12684 non-null  str  
 14  car                   108 non-null    str  
 15  Bar                   12577 non-null  str  
 16  Co

In [5]:
# - duplicates

print("\033[1m" + "\nDuplicated rows:\n" + "\033[0m")

dup = data.duplicated().sum()

print(
    str(dup)
    + " ("
    + str(round(dup / rows * 100, 2))
    + "%)"
)


# Investigate all observations belonging to duplicated sets

duplicates = data[
    data.duplicated(keep=False)
].sort_values(
    by=list(data.columns)
)

print(
    "\033[1m"
    + "\nRows belonging to duplicated sets:\n"
    + "\033[0m"
)

print("Total:", len(duplicates))

duplicates


# Create a temporary dataset without duplicated observations
# for the remaining data-quality investigation

data_quality = data.drop_duplicates().copy()

rows_quality = data_quality.shape[0]

print(
    "\033[1m"
    + "\nRows after removing duplicates:\n"
    + "\033[0m"
)

print(rows_quality)


Duplicated rows:

74 (0.58%)

Rows belonging to duplicated sets:

Total: 148

Rows after removing duplicates:

12610


In [6]:
# - missing counts

print("\033[1m" + "\nMissing data:\n" + "\033[0m")

original_columns = data.columns

missing_data = pd.DataFrame({
    "Total": data_quality[original_columns].isnull().sum(),
    "Percentage": round(
        data_quality[original_columns].isnull().sum()
        / rows_quality * 100, 2
    )
})

print(missing_data)


# - missing percentages - total and by segments

# Automatically identify columns with missing data, excluding 'car'
# because its missingness is assessed separately

cols_with_missing = data_quality.columns[
    data_quality.isnull().any()
    & (data_quality.columns != "car")
]

print(
    "\033[1m"
    + "\nColumns included in missing-data analysis:\n"
    + "\033[0m"
)

for column in cols_with_missing:
    print("•", column)


# Identify rows with at least one missing value
# across the selected columns

data_quality["has_missing"] = (
    data_quality[cols_with_missing]
    .isnull()
    .any(axis=1)
)

print(
    "\033[1m"
    + "\nRows with missing data:\n"
    + "\033[0m"
)

print(
    str(data_quality["has_missing"].sum())
    + " ("
    + str(round(data_quality["has_missing"].mean() * 100, 2))
    + "%)"
)


# Automatically identify variables suitable for segment comparison

segments = [
    column for column in data_quality.columns
    if column not in cols_with_missing
    and column != "car"
    and column != "has_missing"
    and column != "direction_opp"  # redundant with direction_same
    and data_quality[column].nunique() > 1
]


# Check variety of selected segmentation variables

print(
    "\n\033[1m"
    + "Number of unique values by selected segment:"
    + "\033[0m\n"
)

for segment in segments:
    print(
        "•", segment, ":",
        data_quality[segment].nunique()
    )


Missing data:

                      Total  Percentage
destination               0        0.00
passanger                 0        0.00
weather                   0        0.00
temperature               0        0.00
time                      0        0.00
coupon                    0        0.00
expiration                0        0.00
gender                    0        0.00
age                       0        0.00
maritalStatus             0        0.00
has_children              0        0.00
education                 0        0.00
occupation                0        0.00
income                    0        0.00
car                   12502       99.14
Bar                     107        0.85
CoffeeHouse             217        1.72
CarryAway               150        1.19
RestaurantLessThan20    129        1.02
Restaurant20To50        189        1.50
toCoupon_GEQ5min          0        0.00
toCoupon_GEQ15min         0        0.00
toCoupon_GEQ25min         0        0.00
direction_same          

In [7]:
# - calculate missing-data rates for each segment

print("\n\033[1mMissing data by segment\033[0m")

for segment in segments:

    quality = data_quality.groupby(segment)["has_missing"].agg(
        Total="count",
        Missing="sum"
    )

    quality["Missing %"] = round(
        quality["Missing"] / quality["Total"] * 100, 2
    ).astype(str) + "%"

    print("\n\033[1m" + segment + "\033[0m\n")
    print(quality)


Missing data by segment

destination

                 Total  Missing Missing %
destination                              
Home              3230      152     4.71%
No Urgent Place   6266      313      5.0%
Work              3114      138     4.43%

passanger

           Total  Missing Missing %
passanger                          
Alone       7248      334     4.61%
Friend(s)   3292      150     4.56%
Kid(s)      1001       68     6.79%
Partner     1069       51     4.77%

weather

         Total  Missing Missing %
weather                          
Rainy     1202       63     5.24%
Snowy     1397       74      5.3%
Sunny    10011      466     4.65%

temperature

             Total  Missing Missing %
temperature                          
30            2305      121     5.25%
55            3830      178     4.65%
80            6475      304     4.69%

time

      Total  Missing Missing %
time                          
10AM   2271      117     5.15%
10PM   1999      103     5.15%
2PM    2

In [8]:
# - check group sizes for missing-data analysis

# Set the minimum group size used for this diagnostic

min_group_size = 30


# Create lists to collect groups below the minimum group-size criterion

small_total_groups = []
small_complete_groups = []


# Check every group within every segment

for segment in segments:

    quality = data_quality.groupby(segment)["has_missing"].agg(
        Total="count",
        Missing="sum"
    )

    # Calculate observations without missing behavioral data

    quality["Complete"] = (
        quality["Total"] - quality["Missing"]
    )

    # Identify groups with fewer than the minimum number
    # of total observations

    for group in quality.index:

        if quality.loc[group, "Total"] < min_group_size:
            small_total_groups.append(
                (
                    segment,
                    group,
                    quality.loc[group, "Total"],
                    quality.loc[group, "Missing"],
                    quality.loc[group, "Complete"]
                )
            )

        # Separately identify groups with fewer than the minimum
        # number of complete observations

        if quality.loc[group, "Complete"] < min_group_size:
            small_complete_groups.append(
                (
                    segment,
                    group,
                    quality.loc[group, "Total"],
                    quality.loc[group, "Missing"],
                    quality.loc[group, "Complete"]
                )
            )


# Report groups with fewer than the minimum number
# of total observations

print(
    "\n\033[1mGroups with fewer than "
    + str(min_group_size)
    + " total observations:\033[0m"
)

print("Total:", len(small_total_groups))

for segment, group, total, missing, complete in small_total_groups:
    print(
        "•", segment, "-", group,
        "| Total:", total,
        "| Missing:", missing,
        "| Complete:", complete
    )


# Report groups with fewer than the minimum number
# of complete observations

print(
    "\n\033[1mGroups with fewer than "
    + str(min_group_size)
    + " complete observations:\033[0m"
)

print("Total:", len(small_complete_groups))

for segment, group, total, missing, complete in small_complete_groups:
    print(
        "•", segment, "-", group,
        "| Total:", total,
        "| Missing:", missing,
        "| Complete:", complete
    )


Groups with fewer than 30 total observations:
Total: 0

Groups with fewer than 30 complete observations:
Total: 0


### Initial observations

**Dataset size:** The original dataset contains 12,684 observations and 26 variables.

**Duplicated observations:** The dataset contains 74 additional rows that exactly match other observations across all 26 recorded variables (0.58% of the original dataset). These duplicate records were excluded from the subsequent data-quality assessment, leaving 12,610 observations.

**Missing values:** The `car` variable has 99.14% missing values. Five behavioral-frequency variables also contain missing values: `Bar` (0.85%), `CoffeeHouse` (1.72%), `CarryAway` (1.19%), `RestaurantLessThan20` (1.02%), and `Restaurant20To50` (1.50%). Overall, 603 of the 12,610 non-duplicated observations (4.78%) contain at least one missing value across these five behavioral variables.

**Missing-data assessment:** Missingness is relatively consistent across several major segments, including destination, coupon type, gender, coupon acceptance, and direction of travel. However, greater variation is observed across some demographic groups, including age, marital status, education, occupation, income, and whether the respondent has children. Therefore, removing every observation containing a missing behavioral value could change the composition of the analytical sample.

**Group sizes:** All groups included in the exploratory segment-level missing-data assessment contain at least 30 total observations and at least 30 observations without missing behavioral values. The threshold of 30 is used here as an exploratory group-size criterion rather than as evidence of statistical representativeness.

**Other data-quality observations:** Inspection of the categorical variables did not reveal an immediate need for data-type conversion. `toCoupon_GEQ5min` contains only one value and therefore cannot differentiate observations in segment comparisons. `direction_opp` is redundant with `direction_same` because the two variables are exact complements. The misspelled source variable `passanger` is retained unchanged for consistency with the original dataset.

3. Decide what to do about your missing data -- drop, replace, other...

In [11]:
# Remove duplicated observations and the column
# with insufficient data for analysis

data_clean = (
    data
    .drop_duplicates()
    .drop(columns="car")
)

# Check the resulting dataset size

data_clean.shape

(12610, 25)

### Data-cleaning decisions

The 74 additional records that exactly duplicate other observations across all 26 original variables were removed to avoid giving repeated records additional weight in the analysis.

The `car` variable was removed because 99.14% of its values are missing, leaving insufficient observed data for meaningful analysis.

Missing values in `Bar`, `CoffeeHouse`, `CarryAway`, `RestaurantLessThan20`, and `Restaurant20To50` were retained as `NaN`. Although each variable individually has a relatively low proportion of missing values (0.85%–1.72%), removing every observation with at least one missing behavioral value would remove 603 observations, or 4.78% of the non-duplicated dataset. In addition, the proportion of observations with missing behavioral data varies more substantially across some demographic groups, particularly categories of age, marital status, education, occupation, income, and whether the respondent has children. Therefore, complete-case deletion could disproportionately remove observations from some groups and change the composition of the analytical sample.

The exploratory missing-data assessment also showed greater variation in missingness across some demographic groups. Therefore, complete-case deletion could change the composition of the analytical sample.

The missing behavioral values were not imputed because doing so would require unsupported assumptions about unobserved respondent behavior. Instead, observations with missing values will be excluded only from analyses that require a non-missing value for the affected variable.

4. What proportion of the total observations chose to accept the coupon?

5. Use a bar plot to visualize the `coupon` column.

6. Use a histogram to visualize the temperature column.

**Investigating the Bar Coupons**

Now, we will lead you through an exploration of just the bar related coupons.  

1. Create a new `DataFrame` that contains just the bar coupons.


2. What proportion of bar coupons were accepted?


3. Compare the acceptance rate between those who went to a bar 3 or fewer times a month to those who went more.


4. Compare the acceptance rate between drivers who go to a bar more than once a month and are over the age of 25 to the all others.  Is there a difference?


5. Use the same process to compare the acceptance rate between drivers who go to bars more than once a month and had passengers that were not a kid and had occupations other than farming, fishing, or forestry.


6. Compare the acceptance rates between those drivers who:

- go to bars more than once a month, had passengers that were not a kid, and were not widowed *OR*
- go to bars more than once a month and are under the age of 30 *OR*
- go to cheap restaurants more than 4 times a month and income is less than 50K.



7.  Based on these observations, what do you hypothesize about drivers who accepted the bar coupons?

### Independent Investigation

Using the bar coupon example as motivation, you are to explore one of the other coupon groups and try to determine the characteristics of passengers who accept the coupons.  